### PR TITLE
KSP: use compiler's message collector

### PR DESCRIPTION
**Pull request by ting-yuan**
Now that KSP 1.0.3 exits with error message instead of explicit return
code, it needs to inform compiler via compiler's message collector.

Also updated a test case to expect COMPILATION_ERROR instead of
INTERNAL_ERROR, to reflect the behavior change.

### CODE REVIEW

**Comment by ting-yuan** [comment]
KSP 1.0.4 will be out to fix the ABI incompatibility.

**Comment by ting-yuan** [comment]
TL;DR: This binary incompatibility issue isn't related with this PR and the failing test directly. It should be fixed by bumping KSP to 1.0.3, which depends on this PR.

Details about the incompatibility:
This is probably not cleanly fixable in KSP with another release. The issue here is that the new API jar, which is likely a dependency brought by some processor, is not compatible with some old KSP implementation used by KCT. Since there are added APIs, the version of implementation nevertheless need to match it. So @PhilGlass is right, the only fix should be bumping KSP to 1.0.3 in KCT for processors that compile with KSP 1.0.3. (Strictly speaking, for processors that uses new APIs in KSP 1.0.3).

**Comment by ting-yuan** [comment]
KSP 1.0.4 reverted to the old behavior (exit code) by default and only enables the workaround when invoked from Gradle. The tests and abi compatibility should be satisfied now.

Despite of the tests, getting message collector from compiler instead of creating one is more consistent with KSP's implementation and should be less surprising. Therefore the PR could still be useful.

**Comment by ting-yuan** [comment]
Well, the compiler [prioritize](https://github.com/JetBrains/kotlin/blob/ef7b1d1526b4f00c75439c1af4bda8f8093e6183/compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt#L277) error logs over return code so the test will still be broken with this PR.
